### PR TITLE
php: enable ldap support

### LINKED ISSUE #117: Enable PHP LDAP module
The LDAP app is not usable in the snap because the LDAP module is not enabled in php.ini.

The line is already there, just commented out. I'm not sure if this would make the module a dependency or if PHP will gracefully start when a module is not found.

### PR DESCRIPTION
**Pull request by adhami03**
- Enables the LDAP module in PHP, which allows the "LDAP user and group backend" app to be enabled

### CODE REVIEW

File: snap/plugins/x-php.py [hunk 1]
```diff
@@ -99,7 +99,75 @@ def clean_pull(self):
...
+                else: # self.options.configflags
+                    logger.info('No configflags defined???')
+            else: #not libPath == ""
+                logger.info('No libPath defined, cannot fix --with-libdir flag without it. The build may fail.')
+        else: # not distCheck == 0
+            logger.info('No modifications to --with-libdir flag required.')
+
     def build(self):
+        # Check for a broken libpath.
+        self.__fix_library_path()
```
Comment (kyrofa) [inline snap/plugins/x-php.py]: Please help me understand this. This takes the `--with-libdir=` used [here](https://github.com/nextcloud/nextcloud-snap/pull/411/files#diff-5fde7a6d86053f0e1d88c0a2a238941fR167), and turns it into `--with-libdir=/lib/x86_64-linux-gnu` (on amd64, of course)? Why is this required? And it only seems to do this on Ubuntu-- why is it not required elsewhere?

**Comment by kyrofa** [comment]
> How can I apply it to my installation? should I remount fs now RO in RW and apply it?

@mauriziomarini once I'm happy with the code I'll create test snaps that you can use to verify that this works.

File: snap/plugins/x-php.py (hunk 1 above)
Comment (adhami03) [reply snap/plugins/x-php.py]: I believe this is to get around a quirk with Ubuntu, which is mentioned in this comment on the issue: https://github.com/nextcloud/nextcloud-snap/issues/117#issuecomment-256225142 .  I copied the code that @phibbs7 created here (https://github.com/phibbs7/nextcloud-snap/commit/e85418393f407840c5a913990208ae2c91dac0c6) and applied it to the current develop branch.  His fork was out of date and would not build the snap without several updates to the snapcraft.yml

Comment (kyrofa) [reply snap/plugins/x-php.py]: Ah yes, I recall that conversation now.

Snapcraft v2.36 has a way of dealing with this much better, by making `SNAPCRAFT_ARCH_TRIPLET` available for use in the YAML. Only v2.35 is available as a deb right now, though, but we can use that approach as inspiration for a much simpler way to make this work. Then once 2.36 is out it'll be easy to remove this small hack and start using that instead.

**Comment by kyrofa** [comment]
I've actually pushed a new commit here that seems to at least make the app "Enable" button not grayed out. I believe it does the same thing as before, but in a much simpler manner. Thoughts?

I'll spin up snap builds if you agree with the approach.

**Comment by adhami03** [comment]
That code is a lot cleaner and it works the same as before.  I was able to build the snap, install it, enable the LDAP backend, connect it to our AD system and login as an AD user.  Let me know when you get some snaps built that you would like me to test.

**Comment by kyrofa** [comment]
> Let me know when you get some snaps built that you would like me to test.

The builders are currently offline waiting for spectre/meltdown mitigation, I'll spin these as soon as I can.

**Comment by mauriziomarini** [comment]
was merged this pull request 12.0.4?
many thanks

**Comment by adhami03** [comment]
@mauriziomarini , you can test out the changes on my fork here: https://github.com/adhami03/nextcloud-snap .  Just be sure to switch to the develop branch.

**Comment by mauriziomarini** [comment]
@adhami03 Hello Andrew, I am new to snappy, how should I test your branch?

**Comment by adhami03** [comment]
@mauriziomarini , their wiki has a great page about building a snap [here](https://github.com/nextcloud/nextcloud-snap/wiki/Building-your-first-snap).  Follow these steps after you have cloned my repo to your machine and run `git checkout develop` to update to the right branch.

**Comment by kyrofa** [comment]
Sorry guys... the build farm is only starting to come back, I've queued the builds but there seems to be quite a backlog.

**Comment by kyrofa** [comment]
Okay we only have amd64 and i386 builds so far, but they're available in the `stable/pr-411` channel if you're able to test them:

    $ sudo snap install nextcloud --channel=stable/pr-411

Or if you already have it installed:

    $ sudo snap refresh nextcloud --channel=stable/pr-411

**Comment by adhami03** [comment]
@kyrofa , I just installed the amd64 build using your instructions and everything went swimmingly.  Connected to our AD system and was able to log in as domain user.

**Comment by kyrofa** [comment]
Excellent, thank you very much @adhami03. That means this PR is in all likelihood just fine, but I want to make sure things look okay on arm before merging it. Which means we need to wait a little longer for this build farm to come back.

Feel free to continue using the `stable/pr-411` for now, it'll be valid for a month (at which time you'll switch to the stable channel). I suspect this will be in by then.

**Comment by mauriziomarini** [comment]
@kyrofa hello
I did
sudo snap refresh nextcloud --channel=stable/pr-411

Now I can't login as NC compleins:

"You are accessing the server from an untrusted domain.

Please contact your administrator. If you are an administrator of this instance, configure the "trusted_domains" setting in config/config.php. An example configuration is provided in config/config.sample.php.
Depending on your configuration, as an administrator you might also be able to use the button below to trust this domain. "

Sadly, config/config.php  is on readonly fs, so I can;t change it.
What can i do, now?
I am stuck :(

**Comment by haglo** [comment]
Hi thanks for the channel
I have installed it from your channel with
`sudo snap install nextcloud --channel=stable/pr-411
`

when i call the website with http://<Server-IP> 
i get the "Apache2 Ubuntu Default Page"

**Comment by kyrofa** [comment]
@haglo sounds like you have Apache installed from the archives, which is claiming port 80. The snap also wants port 80. You need to either uninstall Apache, or change the ports used by the snap.

**Comment by kyrofa** [comment]
@mauriziomarini you can update trusted domains via `nextcloud.occ`, but you can also edit `/var/snap/nextcloud/current/nextcloud/config/config.php` directly if you prefer.

**Comment by liamdennehy** [comment]
```
user@host:~# uname -rvp && sudo snap info nextcloud | grep tracking
4.4.0-109-generic #132-Ubuntu SMP Tue Jan 9 19:52:39 UTC 2018 x86_64
tracking:                   stable/pr-411
```
Working well. Thanks for activating this, very key feature for my use case, working splendid vs OpenLDAP 2.4

**Comment by kyrofa** [comment]
Hey, alright, the build farm is (mostly) back! There's an armhf version in that channel now, anyone able to test it out on e.g. a raspberry pi? @pachulo, maybe you?

**Comment by pachulo** [comment]
Not right now @kyrofa , but maybe this weekend I can find a moment to try to setup a dummy LDAP server and try it...

**Comment by kyrofa** [comment]
@pachulo even just verifying that you can install the app is enough for me (it should no longer be grayed out).

**Comment by pachulo** [comment]
Well, I installed the build in my rbpi 3 and could activate the application. Now I have this new tab in the admin section:
image

**Comment by kyrofa** [comment]
Sweet, in it goes, thanks @pachulo.
